### PR TITLE
feat(v2): add a way to exclude components from build-time prerendering

### LINKED ISSUE #2273: [v2] add a way to exclude components from SSR within JSX
## 🚀 Feature

Add a way to exclude components from SSR in JSX. Maybe somethin like `<NoSSR></NoSSR>`.

## Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?
yes

## Motivation

Right now, each time we use a react lib that doesn't support SSR we have to put this check `{typeof window !== 'undefined' && .... }` everywhere inside `js` and `mdx files.`

## Pitch

Tools like Next js and nuxt provide a way to exclude components from the SSR within the markup. To avoid errors during the build phase we should be able to do se same in a better way than using `{typeof window !== 'undefined' && .... }` or `{typeof document !== 'undefined' && .... }`.

### PR DESCRIPTION
**Pull request by lex111**
## Motivation

Resolve #2273.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

1. Inside `ClientOnly` put component that used `window` object, eg.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

File: packages/docusaurus/src/client/exports/BrowserOnly.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, {useEffect, useState} from 'react';
+
+const BrowserOnly = ({children}) => {
+  const [isClient, setIsClient] = useState(false);
```
Comment (yangshun) [inline packages/docusaurus/src/client/exports/BrowserOnly.js]: What do you think about renaming it to `isMounted`?

Btw I'm wondering if we can use the `ExecutionEnvironment.canUseDOM` instead, then there's no need for a state right?

Comment (lex111) [reply packages/docusaurus/src/client/exports/BrowserOnly.js]: Why so? This state is needed.

Comment (yangshun) [reply packages/docusaurus/src/client/exports/BrowserOnly.js]: What's the diff between doing your way and this:

```jsx
function BrowserOnly({children}) => {
  return ExecutionEnvironment.canUseDOM && children;
}
```

File: packages/docusaurus/src/client/exports/ClientOnly.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, {useEffect, useState} from 'react';
+
+const ClientOnly = ({children}) => {
+  const [isClient, setIsClient] = useState(false);
```
[the remark this answers is not in the document]
Comment (lex111) [reply packages/docusaurus/src/client/exports/ClientOnly.js]: Done!

File: packages/docusaurus/src/client/exports/BrowserOnly.js (hunk 1 above)
Comment (lex111) [reply packages/docusaurus/src/client/exports/BrowserOnly.js]: It seemed to me that all the work of this component is to use `useEffect`, isn't it?

Comment (yangshun) [reply packages/docusaurus/src/client/exports/BrowserOnly.js]: Yes and I think we don't even need `useEffect` here. Can you try what I've proposed and see whether there's any difference? :/

Comment (lex111) [reply packages/docusaurus/src/client/exports/BrowserOnly.js]: It seems that **none** of the options actually works, it really puzzled me and in complete dismay :face_with_thermometer: 

P.S. For the test, I used the `@u-wave/react-vimeo` component in the MDX file.

Comment (yangshun) [reply packages/docusaurus/src/client/exports/BrowserOnly.js]: So even your current implementation doesn't work? Luckily we test out first lol. Maybe the component is running some side effects that relies on browser environments being present during `import`. In that case there's no way around it.

Comment (lex111) [reply packages/docusaurus/src/client/exports/BrowserOnly.js]: Not exactly, rather why something does not work, this component on some dependencies.

Comment (lex111) [reply packages/docusaurus/src/client/exports/BrowserOnly.js]: After some investigation I found out that build fails if a component using a `window` object is **imported**, how is this possible? Therefore the `BrowserOnly` component is useless here :man_shrugging:

File: packages/docusaurus/src/client/exports/BrowserOnly.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, {useEffect, useState} from 'react';
+
+const BrowserOnly = ({children}) => {
```
Comment (yangshun) [inline packages/docusaurus/src/client/exports/BrowserOnly.js]: This also works and is shorter IMO.

```
function BrowserOnly({children}) {
  return ExecutionEnvironment.canUseDOM && children != null && <>{children()}</>;
};
```

Comment (lex111) [reply packages/docusaurus/src/client/exports/BrowserOnly.js]: Thanks, although this does not solve the issue of importing components using `window`.

For example, if I just import this component in the MDX file, then I will have an error during the building that the `window` is not defined.

```js
import {Replay} from 'vimond-replay';
```

Comment (yangshun) [reply packages/docusaurus/src/client/exports/BrowserOnly.js]: In that case, users will have to lazy load the component after the component has mounted. There's no way we can handle that with the suggested component here.

**Comment by yangshun** [comment]
@lex111 shall we try to get this into the next release?

**Comment by lex111** [comment]
@yangshun done :heavy_check_mark:

File: website/docs/docusaurus-core.md [hunk 4]
```diff
@@ -113,6 +113,18 @@ function Home() {
...
 
+### `<BrowserOnly/>`
+
+This component excludes its `children` during the prerendering stage of the build process. This is useful for hiding code that is only meant to run in the browsers (e.g. where the `window`/`document` objects are being accessed).
+
+```jsx {1,7}
+import BrowserOnly from '@docusaurus/BrowserOnly';
+
+<BrowserOnly>
+  {/* Something that should be excluded during build process prerendering. */}
```
Comment (yangshun) [inline website/docs/docusaurus-core.md]: We need to give an example to show that `children` should be a render function and not normal children.

File: website/docs/docusaurus-core.md [hunk 5]
```diff
@@ -113,6 +113,18 @@ function Home() {
 }
 ```
 
+### `<BrowserOnly/>`
+
+This component excludes its `children` during the prerendering stage of the build process. This is useful for hiding code that is only meant to run in the browsers (e.g. where the `window`/`document` objects are being accessed).
+
+```jsx {1,7}
```
Comment (yangshun) [inline website/docs/docusaurus-core.md]: If we're highlighting the whole block I think it's better we don't highlight at all.

File: website/docs/docusaurus-core.md (hunk 4 above)
Comment (lex111) [reply website/docs/docusaurus-core.md]: Ah yes, done :heavy_check_mark:

**Pull request merged**